src/main/scala/org/broadinstitute/hail/variant/VariantSampleMatrix.scala

```diff
@@ -219,6 +219,23 @@ case class VSMSubgen[T](
   tGen: (Variant) => Gen[T]) {
 
   def gen(sc: SparkContext)(implicit tct: ClassTag[T]): Gen[VariantSampleMatrix[T]] =
+    this.distributeSize.flatMap(x => x.genNoResize(sc))
+
+  def distributeSize: Gen[VSMSubgen[T]] = Gen { (p: Parameters) =>
+    val subsizes: Array[Int] = Gen.partition(p.rng, p.size, 9)
+    VSMSubgen(sampleIdGen.resize(subsizes(0)),
+      saSigGen.resize(subsizes(1)),
+      vaSigGen.resize(subsizes(2)),
+      globalSigGen.resize(subsizes(3)),
+      (i,t) => saGen(i, t).resize(subsizes(4)),
+      (t) => vaGen(t).resize(subsizes(5)),
+      (t) => globalGen(t).resize(subsizes(6)),
+      vGen.resize(subsizes(7)),
+      (v) => tGen(v).resize(subsizes(8))
+    )
```

Comment:
Great start!  I think this approach is still worse-case O(n^2) since each subsize is worst case O(n / 9) = O(n) and we're constructing a matrix.  A few more remarks:

Variants and gentypes (t) are O(1) since the number of alternate alleles are bounded, see Variant.gen:

```
nAlleles <- Gen.frequency((5, Gen.const(2)), (1, Gen.choose(2, 10)))
```

However, if this was used to generate a VSM with a different T, it could potentially be cubic.

The number of variants (rows) isn't resized in this patch since VSMSubgen.gen calls Gen.distinctBuildableOf directly.

I think the total generated size is:  O(nSamples*nVariants*size(t) + nSamples*max(size(saSig), size(sa)) + nVariants*max(size(vaSig), size(va)) + max(size(globalSig), size(global)), where nSamples is the size of generated by sampleIdsGen and nVariants is the size generated by the distinctBuildableOf row generator.